### PR TITLE
executor, util: vectorize hash calculation during probing

### LINKED ISSUE #12048: Vectorize hash calculation in hashJoin.
## Feature Request

**Is your feature request related to a problem? Please describe:**

In `hashJoin`, we do the hash calculation for a chunk row by row https://github.com/pingcap/tidb/blob/1ff620d/executor/hash_table.go#L143-L153, can try to vectorize it for performance.

**Describe the feature you'd like:**

- [x] vectorize hash calculation during building hash table. #12076 by @sduzh 
- [x] vectorize hash calculation during probing. #12669 by @sduzh 

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by sduzh**
### What problem does this PR solve? 
Fix #12048 

### What is changed and how it works?
```
name                                                                       old time/op  new time/op  delta
HashJoinExec/(rows:100000,_concurency:4,_joinKeyIdx:_[0_1],_disk:false)-8   716ms ± 1%   717ms ± 3%    ~     (p=1.000 n=16+20)
HashJoinExec/(rows:100000,_concurency:4,_joinKeyIdx:_[0],_disk:false)-8     102ms ± 2%   100ms ± 1%  -1.16%  (p=0.000 n=20+18)
HashJoinExec/(rows:100000,_concurency:4,_joinKeyIdx:_[0],_disk:true)-8      708ms ±13%   700ms ±11%    ~     (p=0.640 n=20+20)
HashJoinExec/(rows:1000,_concurency:4,_joinKeyIdx:_[0],_disk:true)-8       17.8ms ±20%  17.9ms ±15%    ~     (p=0.968 n=20+20)
```

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

 - Possible performance regression
 - Increased code complexity

Related changes

Release note

 - Write release note for bug-fix or new feature.

### CODE REVIEW

File: executor/join.go [hunk 1]
```diff
@@ -419,11 +420,22 @@ func (e *HashJoinExec) join2Chunk(workerID uint, outerChk *chunk.Chunk, hCtx *ha
 		joinResult.err = err
 		return false, joinResult
 	}
+
+	hCtx.initHash(outerChk.NumRows())
+	for _, i := range hCtx.keyColIdx {
+		err = codec.HashChunkSelected(e.rowContainer.sc, hCtx.hashVals, outerChk, hCtx.allTypes[i], i, hCtx.buf, hCtx.hasNull, selected)
```
Comment (SunRunAway) [inline executor/join.go]:
```suggestion
		err = codec.HashChunkSelected(e.ctx, hCtx.hashVals, outerChk, hCtx.allTypes[i], i, hCtx.buf, hCtx.hasNull, selected)
```

**Comment by SunRunAway** [comment]
Hi, @sduzh
Thank you for your PR, your code looks great.

May you help do a little more profiling?

Use `go test -run=XX -bench="BenchmarkHashJoinExec" -benchmem -count 5  -memprofile memprofile.out -cpuprofile profile.out` to collect `profile.out`

`go tool pprof -http=":8080" profile.out` to analyze through flame graph, or use `go tool pprof profile.out` and `list` command in CLI to analyze the cumulative cost of hash calculation during probing. Maybe you would like to change the function name of `getJoinKeyFromChkRow` for probing to get the cumulative cost during probing in old code.

Also, you may just target this test `HashJoinExec/(rows:100000,_concurency:4,_joinKeyIdx:_[0],_disk:false)-8`, and temporarily disable the others.

File: util/codec/codec.go [hunk 2]
```diff
@@ -360,13 +360,24 @@ func encodeHashChunkRowIdx(sc *stmtctx.StatementContext, row chunk.Row, tp *type
...
+// sel indicates which rows are selected. If it is nil, all rows are selected.
+func HashChunkSelected(sc *stmtctx.StatementContext, h []hash.Hash64, chk *chunk.Chunk, tp *types.FieldType, colIdx int, buf []byte,
+	isNull, sel []bool) (err error) {
 	var b []byte
 	column := chk.Column(colIdx)
 	rows := chk.NumRows()
 	switch tp.Tp {
 	case mysql.TypeTiny, mysql.TypeShort, mysql.TypeInt24, mysql.TypeLong, mysql.TypeLonglong, mysql.TypeYear:
 		i64s := column.Int64s()
+		column.MergeNulls()
```
Comment (SunRunAway) [inline util/codec/codec.go]: Is it added by accident?

**Comment by sduzh** [comment]
> Hi, @sduzh
> Thank you for your PR, your code looks great.
> 
> May you help do a little more profiling?
> 
> Use `go test -run=XX -bench="BenchmarkHashJoinExec" -benchmem -count 5 -memprofile memprofile.out -cpuprofile profile.out` to collect `profile.out`
> 
> `go tool pprof -http=":8080" profile.out` to analyze through flame graph, or use `go tool pprof profile.out` and `list` command in CLI to analyze the cumulative cost of hash calculation during probing. Maybe you would like to change the function name of `getJoinKeyFromChkRow` for probing to get the cumulative cost during probing in old code.
> 
> Also, you may just target this test `HashJoinExec/(rows:100000,_concurency:4,_joinKeyIdx:_[0],_disk:false)-8`, and temporarily disable the others.

Thanks for your advice, I will try it tonight.

File: util/codec/codec.go (hunk 2 above)
Comment (sduzh) [reply util/codec/codec.go]: Yes

**Comment by sduzh** [comment]
@SunRunAway 
results on this new branch
```
➜  executor git:(vec-hash-probe) ✗ go tool pprof profile.out
Type: cpu
Time: Oct 18, 2019 at 1:27am (CST)
Duration: 23.05s, Total samples = 38.50s (167.05%)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) list join2Chunk
Total: 38.50s
ROUTINE ======================== github.com/pingcap/tidb/executor.(*HashJoinExec).join2Chunk in /Users/zhuming/go/src/github.com/sduzh/tidb/executor/join.go
     130ms     10.07s (flat, cum) 26.16% of Total
         .          .    421:		return false, joinResult
         .          .    422:	}
         .          .    423:
         .          .    424:	hCtx.initHash(outerChk.NumRows())
         .          .    425:	for _, i := range hCtx.keyColIdx {
         .       20ms    426:		err = codec.HashChunkSelected(e.rowContainer.sc, hCtx.hashVals, outerChk, hCtx.allTypes[i], i, hCtx.buf, hCtx.hasNull, selected)
         .          .    427:		if err != nil {
         .          .    428:			joinResult.err = err
         .          .    429:			return false, joinResult
         .          .    430:		}
         .          .    431:	}
         .          .    432:
      20ms       20ms    433:	for i := range selected {
      50ms       50ms    434:		if !selected[i] || hCtx.hasNull[i] { // process unmatched outer rows
         .          .    435:			e.joiners[workerID].onMissMatch(false, outerChk.GetRow(i), joinResult.chk)
         .          .    436:		} else { // process matched outer rows
      30ms       50ms    437:			probeKey, probeRow := hCtx.hashVals[i].Sum64(), outerChk.GetRow(i)
      30ms      9.93s    438:			ok, joinResult = e.joinMatchedOuterRow2Chunk(workerID, probeKey, probeRow, hCtx, joinResult)
         .          .    439:			if !ok {
         .          .    440:				return false, joinResult
         .          .    441:			}
         .          .    442:		}
         .          .    443:		if joinResult.chk.IsFull() {
```

results on master branch
```
➜  executor git:(master) ✗ go tool pprof profile.out.old
Type: cpu
Time: Oct 18, 2019 at 1:34am (CST)
Duration: 20.49s, Total samples = 33.69s (164.45%)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) list join2Chunk
Total: 33.69s
ROUTINE ======================== github.com/pingcap/tidb/executor.(*HashJoinExec).join2Chunk in /Users/zhuming/go/src/github.com/sduzh/tidb/executor/join.go
      30ms      9.85s (flat, cum) 29.24% of Total
         .          .    412:}
         .          .    413:
         .          .    414:func (e *HashJoinExec) join2Chunk(workerID uint, outerChk *chunk.Chunk, hCtx *hashContext, joinResult *hashjoinWorkerResult,
         .          .    415:	selected []bool) (ok bool, _ *hashjoinWorkerResult) {
         .          .    416:	var err error
      10ms       20ms    417:	selected, err = expression.VectorizedFilter(e.ctx, e.outerFilter, chunk.NewIterator4Chunk(outerChk), selected)
         .          .    418:	if err != nil {
         .          .    419:		joinResult.err = err
         .          .    420:		return false, joinResult
         .          .    421:	}
         .          .    422:	for i := range selected {
      10ms       10ms    423:		if !selected[i] { // process unmatched outer rows
         .          .    424:			e.joiners[workerID].onMissMatch(false, outerChk.GetRow(i), joinResult.chk)
         .          .    425:		} else { // process matched outer rows
      10ms      9.82s    426:			ok, joinResult = e.joinMatchedOuterRow2Chunk(workerID, outerChk.GetRow(i), hCtx, joinResult)
         .          .    427:			if !ok {
         .          .    428:				return false, joinResult
         .          .    429:			}
         .          .    430:		}
         .          .    431:		if joinResult.chk.IsFull() {
(pprof) list getJoinKeyFromChkRow
Total: 33.69s
ROUTINE ======================== github.com/pingcap/tidb/executor.(*hashRowContainer).getJoinKeyFromChkRowX in /Users/zhuming/go/src/github.com/sduzh/tidb/executor/hash_table.go
      30ms      290ms (flat, cum)  0.86% of Total
         .          .    239:}
         .          .    240:
         .          .    241:// getJoinKeyFromChkRow fetches join keys from row and calculate the hash value.
         .          .    242:func (*hashRowContainer) getJoinKeyFromChkRow(sc *stmtctx.StatementContext, row chunk.Row, hCtx *hashContext) (hasNull bool, key uint64, err error) {
         .          .    243:	for _, i := range hCtx.keyColIdx {
         .       10ms    244:		if row.IsNull(i) {
         .          .    245:			return true, 0, nil
         .          .    246:		}
         .          .    247:	}
      10ms       50ms    248:	hCtx.initHash(1)
      10ms      220ms    249:	err = codec.HashChunkRow(sc, hCtx.hashVals[0], row, hCtx.allTypes, hCtx.keyColIdx, hCtx.buf)
      10ms       10ms    250:	return false, hCtx.hashVals[0].Sum64(), err
         .          .    251:}
         .          .    252:
         .          .    253:// Len returns the length of the records in hashRowContainer.
         .          .    254:func (c hashRowContainer) Len() int {
         .          .    255:	return c.hashTable.Len()
(pprof)
```

**Comment by sduzh** [comment]
> @SunRunAway
> results on this new branch
> 
> ```
> ➜  executor git:(vec-hash-probe) ✗ go tool pprof profile.out
> Type: cpu
> Time: Oct 18, 2019 at 1:27am (CST)
> Duration: 23.05s, Total samples = 38.50s (167.05%)
> Entering interactive mode (type "help" for commands, "o" for options)
> (pprof) list join2Chunk
> Total: 38.50s
> ROUTINE ======================== github.com/pingcap/tidb/executor.(*HashJoinExec).join2Chunk in /Users/zhuming/go/src/github.com/sduzh/tidb/executor/join.go
>      130ms     10.07s (flat, cum) 26.16% of Total
>          .          .    421:		return false, joinResult
>          .          .    422:	}
>          .          .    423:
>          .          .    424:	hCtx.initHash(outerChk.NumRows())
>          .          .    425:	for _, i := range hCtx.keyColIdx {
>          .       20ms    426:		err = codec.HashChunkSelected(e.rowContainer.sc, hCtx.hashVals, outerChk, hCtx.allTypes[i], i, hCtx.buf, hCtx.hasNull, selected)
>          .          .    427:		if err != nil {
>          .          .    428:			joinResult.err = err
>          .          .    429:			return false, joinResult
>          .          .    430:		}
>          .          .    431:	}
>          .          .    432:
>       20ms       20ms    433:	for i := range selected {
>       50ms       50ms    434:		if !selected[i] || hCtx.hasNull[i] { // process unmatched outer rows
>          .          .    435:			e.joiners[workerID].onMissMatch(false, outerChk.GetRow(i), joinResult.chk)
>          .          .    436:		} else { // process matched outer rows
>       30ms       50ms    437:			probeKey, probeRow := hCtx.hashVals[i].Sum64(), outerChk.GetRow(i)
>       30ms      9.93s    438:			ok, joinResult = e.joinMatchedOuterRow2Chunk(workerID, probeKey, probeRow, hCtx, joinResult)
>          .          .    439:			if !ok {
>          .          .    440:				return false, joinResult
>          .          .    441:			}
>          .          .    442:		}
>          .          .    443:		if joinResult.chk.IsFull() {
> ```
> 
> results on master branch
> 
> ```
> ➜  executor git:(master) ✗ go tool pprof profile.out.old
> Type: cpu
> Time: Oct 18, 2019 at 1:34am (CST)
> Duration: 20.49s, Total samples = 33.69s (164.45%)
> Entering interactive mode (type "help" for commands, "o" for options)
> (pprof) list join2Chunk
> Total: 33.69s
> ROUTINE ======================== github.com/pingcap/tidb/executor.(*HashJoinExec).join2Chunk in /Users/zhuming/go/src/github.com/sduzh/tidb/executor/join.go
>       30ms      9.85s (flat, cum) 29.24% of Total
>          .          .    412:}
>          .          .    413:
>          .          .    414:func (e *HashJoinExec) join2Chunk(workerID uint, outerChk *chunk.Chunk, hCtx *hashContext, joinResult *hashjoinWorkerResult,
>          .          .    415:	selected []bool) (ok bool, _ *hashjoinWorkerResult) {
>          .          .    416:	var err error
>       10ms       20ms    417:	selected, err = expression.VectorizedFilter(e.ctx, e.outerFilter, chunk.NewIterator4Chunk(outerChk), selected)
>          .          .    418:	if err != nil {
>          .          .    419:		joinResult.err = err
>          .          .    420:		return false, joinResult
>          .          .    421:	}
>          .          .    422:	for i := range selected {
>       10ms       10ms    423:		if !selected[i] { // process unmatched outer rows
>          .          .    424:			e.joiners[workerID].onMissMatch(false, outerChk.GetRow(i), joinResult.chk)
>          .          .    425:		} else { // process matched outer rows
>       10ms      9.82s    426:			ok, joinResult = e.joinMatchedOuterRow2Chunk(workerID, outerChk.GetRow(i), hCtx, joinResult)
>          .          .    427:			if !ok {
>          .          .    428:				return false, joinResult
>          .          .    429:			}
>          .          .    430:		}
>          .          .    431:		if joinResult.chk.IsFull() {
> (pprof) list getJoinKeyFromChkRow
> Total: 33.69s
> ROUTINE ======================== github.com/pingcap/tidb/executor.(*hashRowContainer).getJoinKeyFromChkRowX in /Users/zhuming/go/src/github.com/sduzh/tidb/executor/hash_table.go
>       30ms      290ms (flat, cum)  0.86% of Total
>          .          .    239:}
>          .          .    240:
>          .          .    241:// getJoinKeyFromChkRow fetches join keys from row and calculate the hash value.
>          .          .    242:func (*hashRowContainer) getJoinKeyFromChkRow(sc *stmtctx.StatementContext, row chunk.Row, hCtx *hashContext) (hasNull bool, key uint64, err error) {
>          .          .    243:	for _, i := range hCtx.keyColIdx {
>          .       10ms    244:		if row.IsNull(i) {
>          .          .    245:			return true, 0, nil
>          .          .    246:		}
>          .          .    247:	}
>       10ms       50ms    248:	hCtx.initHash(1)
>       10ms      220ms    249:	err = codec.HashChunkRow(sc, hCtx.hashVals[0], row, hCtx.allTypes, hCtx.keyColIdx, hCtx.buf)
>       10ms       10ms    250:	return false, hCtx.hashVals[0].Sum64(), err
>          .          .    251:}
>          .          .    252:
>          .          .    253:// Len returns the length of the records in hashRowContainer.
>          .          .    254:func (c hashRowContainer) Len() int {
>          .          .    255:	return c.hashTable.Len()
> (pprof)
> ```

It looks like the cost of hash calculation tasks only a small fraction of the total cost of the hash probing phase, less than 1% in both branches.

File: executor/join.go (hunk 1 above)
Comment (sduzh) [reply executor/join.go]: The implementation of `HashChunkColumns` relies on `HashChunkSelected`.
It is inconvenient to pass `sessionctx.Context` from `HashChunkColumns`.

**Comment by SunRunAway** [comment]
Sorry for the late reply, I'm really busy in the last week.

And thanks for your work, the result and conclusion you post is really helpful.
I think it's the problem of the test case.
The test case must join two tables which both have wide columns (see https://github.com/pingcap/tidb/blob/cc991d90d39bfad7bd8b76575d7a488cdf1800ac/executor/benchmark_test.go#L542).
Could you try to design a new test case?

**Comment by sduzh** [comment]
> Sorry for the late reply, I'm really busy in last week.
> 
> And thanks for your work, the result and conclusion you post is really helpful.
> I think it's the problem of the test case.
> The test case must join two tables which both have wide columns (see
> 
> https://github.com/pingcap/tidb/blob/cc991d90d39bfad7bd8b76575d7a488cdf1800ac/executor/benchmark_test.go#L542
> 
> 
> The column of index 1 is a string type, which the size of it is 5K).
> Could you try to design a new test case?

No problem.

**Comment by XuHuaiyu** [comment]
Friendly ping, is this PR still work in process? @sduzh

**Comment by sduzh** [comment]
Sorry for the late reply. 
As @SunRunAway mentioned above, the original test case must join the wide string columns, which result in a lot of string copies.
So I designed a new test case, as suggested by @SunRunAway, replace the string columns with double columns, and the result is much better:
```
➜  executor git:(vec-hash-probe) ✗ benchstat old.txt new.txt
name                                                                                                        old time/op  new time/op  delta
HashJoinExec/(rows:100000,_cols:[[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0_1],_disk:false)-8         28.7ms ± 1%  23.5ms ± 1%  -18.10%  (p=0.008 n=5+5)
HashJoinExec/(rows:100000,_cols:[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0],_disk:false)-8           26.2ms ± 3%  22.8ms ± 4%  -12.77%  (p=0.008 n=5+5)

``` 
@XuHuaiyu

**Comment by SunRunAway** [comment]
> Sorry for the late reply.
> As @SunRunAway mentioned above, the original test case must join the wide string columns, which result in a lot of string copies.
> So I designed a new test case, as suggested by @SunRunAway, replace the string columns with double columns, and the result is much better:
> 
> ```
> ➜  executor git:(vec-hash-probe) ✗ benchstat old.txt new.txt
> name                                                                                                        old time/op  new time/op  delta
> HashJoinExec/(rows:100000,_cols:[[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0_1],_disk:false)-8         28.7ms ± 1%  23.5ms ± 1%  -18.10%  (p=0.008 n=5+5)
> HashJoinExec/(rows:100000,_cols:[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0],_disk:false)-8           26.2ms ± 3%  22.8ms ± 4%  -12.77%  (p=0.008 n=5+5)
> ```
> 
> @XuHuaiyu

Great job, you can post the newest benchmark result into the PR description.

**Comment by XuHuaiyu** [comment]
> 04.sql duration: 8190.27 ms ± 2.16% (std=177.18) delta: 4.82% (p=0.237)

Is this expected?

**Comment by sduzh** [comment]
> > 04.sql duration: 8190.27 ms ± 2.16% (std=177.18) delta: 4.82% (p=0.237)
> 
> Is this expected?

No

**Comment by SunRunAway** [comment]
Thank you, @sduzh.

**Pull request merged**